File: Next Greater Element III/string.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

int solution(int num) {
    string s = to_string(num);
    
    int i, j = i = s.size() - 1;
    while (i > 0 && s[i-1] >= s[i]) i--; 
    if (i == 0) return -1;
    
    while (j >= i && s[j] <= s[i-1]) j--;
    swap(s[i-1], s[j]);
    
    reverse(s.begin()+i, s.end());
    return stoll(s) <= INT32_MAX ? stoll(s) : -1;
}

int main() {
    int num = 1234765553;
    cout << solution(num) << endl;
    num = 765553;
    cout << solution(num) << endl;
    num = 1999999999;
    cout << solution(num) << endl;
}

File: Next Greater Element III/string.py
def solution(num):
    num = list(str(num))
    i = j = len(num) - 1
    while i > 0:
        if num[i-1] < num[i]: break
        i -= 1
    
    if i == 0: return -1
    
    while j < len(num) and j >= i:
        if num[j] > num[i-1]: break
        j -= 1
    
    num[i-1], num[j] = num[j], num[i-1]
    ans = int("".join(num[:i] + num[i:][::-1]))
    return ans if ans <= 2**31 else -1

num = 1234765553
solution(num)
num = 765553
solution(num)
num = 1999999999
solution(num)
